api/envoy/extensions/transport_socket/capture/v2/capture.proto

```diff
@@ -0,0 +1,47 @@
+syntax = "proto3";
+
+// Trace capture format for the capture transport socket extension. This dumps plain text read/write
+// sequences on a socket.
+
+package envoy.extensions.transport_socket.capture.v2;
+option go_package = "v2";
+
+import "envoy/api/v2/core/address.proto";
+
+import "google/protobuf/timestamp.proto";
+
+// Connection properties.
+message Connection {
+  // Global unique connection ID for Envoy session. Matches connection IDs used
+  // in Envoy logs.
+  uint64 id = 1;
+  // Local address.
+  envoy.api.v2.core.Address local_address = 2;
+  // Remote address.
+  envoy.api.v2.core.Address remote_address = 3;
+}
+
+// Event in a capture trace.
+message Event {
+  // Timestamp for event.
+  google.protobuf.Timestamp timestamp = 1;
+  message Read {
+    bytes data = 1;
+  }
+  message Write {
+    bytes data = 1;
+  }
+  // Read or write with content as bytes string.
+  oneof event_selector {
+    Read read = 2;
+    Write write = 3;
+  }
+}
+
+// Sequence of read/write events that constitute a captured trace on a socket.
+message Trace {
```

Comment:
For long traces, I don't think we will want to store all events in a single message. There probably needs to be some limit to the number of events per trace message, and then depending on the output format there will be multiple trace records (or for API target they would potentially just be streamed.). Can you think about that a bit and then document at least what the record layout would be for a file?

Reply:
I think you can just dump multiple `Trace` down the same gRPC stream (for example). You would have the connection ID in each, so they can be reassembled at the other side. You'd only need to include other connection properties like remote/local address in the first message sent.

Reply:
Yup agreed. For files though I think some framing might be required but I'm not sure. I don't have much experience writing protos to files.

Reply:
For files, I don't think there is a great solution, see the discussion on large data sets at https://developers.google.com/protocol-buffers/docs/techniques. I will add a TODO to consider how to incrementally stream out trace fragments rather than waiting until the end to dump, I think what we have there right now is a reasonable start for this PR though.

Reply:
Yeah I think for files it will need to be a stream of length delimited messages. TODO is fine for now.